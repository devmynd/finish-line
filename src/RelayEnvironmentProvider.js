import { Component } from 'react'
import PropTypes from 'prop-types'
import { commitMutation } from 'react-relay'

export class RelayEnvironmentProvider extends Component {
  static propTypes = {
    create: PropTypes.func.isRequired,
    children: PropTypes.any.isRequired
  }

  static childContextTypes = {
    relayEnvironment: PropTypes.shape({
      commitMutation: PropTypes.func.isRequired,
      current: PropTypes.any.isRequired,
      refresh: PropTypes.func.isRequired
    }).isRequired
  }

  state = {
    environment: this.props.create()
  }

  getChildContext () {
    return {
      relayEnvironment: {
        commitMutation: this.commitMutation,
        current: this.state.environment,
        refresh: this.refresh
      }
    }
  }

  commitMutation = (arg) => {
    commitMutation(this.state.environment, arg)
  }

  refresh = (callback) => {
    const environment = this.props.create()
    this.setState({ environment }, callback)
  }

  render () {
    return this.props.children
  }
}
