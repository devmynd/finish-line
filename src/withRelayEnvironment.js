import React, { Component } from 'react'
import PropTypes from 'prop-types'
import hoistNonReactStatics from 'hoist-non-react-statics'
import { RelayEnvironment } from './RelayEnvironment'

export const withRelayEnvironment = (GivenComponent) => {
  class NewComponent extends Component {
    static propTypes = {
      wrappedComponentRef: PropTypes.func
    }

    render () {
      const { wrappedComponentRef, ...otherProps } = this.props

      return <RelayEnvironment>
        {relayEnvironment =>
          <GivenComponent ref={wrappedComponentRef} relayEnvironment={relayEnvironment} {...otherProps} />
        }
      </RelayEnvironment>
    }
  }

  NewComponent.displayName = `withRelayEnvironment(${GivenComponent.displayName || GivenComponent.name})`
  NewComponent.WrappedComponent = GivenComponent

  return hoistNonReactStatics(NewComponent, GivenComponent)
}
