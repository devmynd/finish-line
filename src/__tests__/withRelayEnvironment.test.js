/* eslint-env jest */
import React, { Component } from 'react'
import { mount } from 'enzyme'
import { uniqueId } from 'lodash'
import { RelayEnvironmentProvider } from '../RelayEnvironmentProvider'
import { withRelayEnvironment } from '../withRelayEnvironment'

describe('provided props', () => {
  let givenProps
  let envProvider

  const Dummy = withRelayEnvironment((props) => {
    givenProps = props
    return <div />
  })

  beforeEach(() => {
    envProvider = mount(<RelayEnvironmentProvider environmentProvider={uniqueId}>
      <Dummy />
    </RelayEnvironmentProvider>)
  })

  it('is the context from the environment provider', () => {
    expect(givenProps).toEqual(envProvider.instance().getChildContext())
  })
})

describe('displayName', () => {
  it('is the existing display name wrapped', () => {
    const component = () => null
    component.displayName = 'display-name'
    const wrapped = withRelayEnvironment(component)
    expect(wrapped.displayName).toEqual(`withRelayEnvironment(${component.displayName})`)
  })

  it('is the name wrapped if there is no existing display name', () => {
    const component = () => null
    component.displayName = null
    const wrapped = withRelayEnvironment(component)
    expect(wrapped.displayName).toEqual(`withRelayEnvironment(${component.name})`)
  })
})

describe('statics', () => {
  it('hoists the statics', () => {
    const Dummy = () => null
    Dummy.foo = 'foo'
    Dummy.bar = () => 'bar'
    const wrapped = withRelayEnvironment(Dummy)
    expect(wrapped.foo).toEqual(Dummy.foo)
    expect(wrapped.bar).toEqual(Dummy.bar)
  })
})

describe('WrappedComponent', () => {
  it('is the wrapped component', () => {
    const Dummy = () => null
    const wrapped = withRelayEnvironment(Dummy)
    expect(wrapped.WrappedComponent).toEqual(Dummy)
  })
})

describe('wrappedComponentRef', () => {
  it('does not pass the wrapped component ref to the wrapped component', () => {
    class Dummy extends Component {
      render () { return null }
    }
    const Wrapped = withRelayEnvironment(Dummy)
    const subject = mount(<RelayEnvironmentProvider environmentProvider={() => 1}>
      <Wrapped wrappedComponentRef={() => 'hi'} />
    </RelayEnvironmentProvider>)
    expect(subject).toMatchSnapshot()
  })
})
