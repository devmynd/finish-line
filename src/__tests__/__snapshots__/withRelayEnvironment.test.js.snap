// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrappedComponentRef does not pass the wrapped component ref to the wrapped component 1`] = `
<RelayEnvironmentProvider
  create={[Function]}
>
  <withRelayEnvironment(Dummy)
    wrappedComponentRef={[Function]}
  >
    <Dummy
      relayEnvironment={
        Object {
          "commitMutation": [Function],
          "current": 1,
          "refresh": [Function],
        }
      }
    />
  </withRelayEnvironment(Dummy)>
</RelayEnvironmentProvider>
`;
